mantidproject/mantid
# Improve speed of empty instrument processing in PEARL powder diffraction scripts

**Description of work.**

The speed of the processing of empty instrument workspaces has been improved in the ISIS powder diffraction scripts. This has been achieved by summing the empty workspaces during the (infrequent) calibration process and then loading the presummed file during the (frequent) Focus processes

This behaviour was requested by PEARL because they have a larger number of empty instrument workspaces than the other ISIS powder diffraction instruments. Nick Funnell has checked the other ISIS instruments are also happy to go with this new functionality.

I have also exposed the sample_empty and sample_empty_scale parameters in the PEARL configuration. These were previously available in the Polaris and GEM versions of the scripts. These parameters allow the empty run set specified in the calibration config to be overridden during a Focus process

To provide some backwards compatibility, if a Focus process is run and no pre-summed empty instrument workspace file is found then the Focus process will generate its own sum (previous behaviour)

**Report to:** Nick Funnell

**To test:**

Unzip the attached zip file into a folder.
Adjust paths in the test.py file to point at the folder you've just created
Also adjust path in the user_experiment.yaml file
Run test.py from Workbench (will need archive access enabled).
This will run a Calibration followed by a Focus - each step will take a few mins. It might be easier to observe the following test results if you run the Calibration first and then the Focus

1. Observe that a file called "summed_empty_95648-95654.nxs" has been written by the calibration process to a subfolder called calib_critical_files
2. Observe that the focus process loads summed_empty_95648-95654.nxs instead of loading the workspaces 95648 to 95654 individually and summing them
3. Rename the "summed_empty_95648-95654.nxs" file to something else eg summed_empty_95648-95654.nxstest
4. Rerun the focus step and observe that it still works but takes longer and the workspaces 95648 to 95654 are loaded

[PearlFocusEmptyInstTest.zip](https://github.com/mantidproject/mantid/files/5366245/PearlFocusEmptyInstTest.zip)

Fixes #29478. 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
